src/utils/typey-type.ts

```diff
@@ -0,0 +1,188 @@
+import type { PersonalDictionaryNameAndContents } from "types";
+
+type LocalStoragePersonalDictionariesV0 = PersonalDictionaryNameAndContents[];
+
+type LocalStoragePersonalDictionariesV1 = {
+  v: string;
+  dicts: PersonalDictionaryNameAndContents[];
+};
+
+type LocalStoragePersonalDictionariesUnknownVersion =
+  | LocalStoragePersonalDictionariesV0
+  | LocalStoragePersonalDictionariesV1;
+
+function isV1(
+  data: LocalStoragePersonalDictionariesV0 | LocalStoragePersonalDictionariesV1
+): data is LocalStoragePersonalDictionariesV1 {
+  return "v" in data;
+}
+
+const migratePersonalDictionariesV0ToV1 = function(
+  localStoragePersonalDictionaries: LocalStoragePersonalDictionariesUnknownVersion,
+  dirtyFlag: boolean
+): [LocalStoragePersonalDictionariesV1, boolean] {
+  if (isV1(localStoragePersonalDictionaries)) {
+    return [localStoragePersonalDictionaries, dirtyFlag];
+  } else {
+    return [
+      {
+        v: "1",
+        dicts: localStoragePersonalDictionaries,
+      },
+      true,
+    ];
+  }
+};
+
+// const migratePersonalDictionariesV1ToV2 = function (personalDictionaries, dirtyFlag) {
+//   if (personalDictionaries.v && personalDictionaries.v === '1') {
+//     let opts = {};
+//     let dictsWithMetadata = personalDictionaries.dicts.map(dict => [dict[0],dict[1],opts]);
+//     personalDictionaries = {v:'2',dicts:dictsWithMetadata};
+//     dirtyFlag = true;
+//   }
+
+//   return [personalDictionaries, dirtyFlag];
+// }
+
+const runAllPersonalDictionariesMigrations = function(
+  personalDictionaries: LocalStoragePersonalDictionariesUnknownVersion,
+  dirtyFlag: boolean
+): [LocalStoragePersonalDictionariesV1 | null, boolean, string | null] {
+  let error = null;
+  try {
+    [personalDictionaries, dirtyFlag] = migratePersonalDictionariesV0ToV1(
+      personalDictionaries,
+      dirtyFlag
+    );
+    // [personalDictionaries, dirtyFlag] = migratePersonalDictionariesV1ToV2(personalDictionaries, dirtyFlag);
+  } catch (exception) {
+    return [null, false, "Personal dictionaries found in local storage are either missing a version number or are not in valid version 0 format for migrating to version 1."];
+  }
+  return [personalDictionaries, dirtyFlag, error];
+};
+
+function migratePersonalDictionariesV(personalDictionaries: LocalStoragePersonalDictionariesUnknownVersion): [LocalStoragePersonalDictionariesV1 | null, string | null] {
+  let dirtyFlag = false;
+  let error: string | null = null;
+  let maybePersonalDictionaries: null | LocalStoragePersonalDictionariesUnknownVersion = personalDictionaries;
```

Comment:
Adds `let maybePersonalDictionaries: null | LocalStoragePersonalDictionariesUnknownVersion = personalDictionaries;` instead of re-assigning `personalDictionaries`.